sklearn/ensemble/_hist_gradient_boosting/tests/test_binning.py

```diff
@@ -312,3 +317,140 @@ def test_infinite_values():
 
     expected_binned_X = np.array([0, 1, 2, 3]).reshape(-1, 1)
     assert_array_equal(bin_mapper.transform(X), expected_binned_X)
+
+
+@pytest.mark.parametrize("categorical", [None, np.array([False])])
+def test_categorical_transform_error_with_no_categoricals(categorical):
+    # Properties and
+    X = np.arange(10, dtype=float).reshape(-1, 1)
+    bin_mapper = _BinMapper(n_bins=5, categorical=categorical).fit(X)
+
+    assert len(bin_mapper.bin_categories_) == 0
+
+    msg = ("categorical_only=True can only be set when there are "
+           "categorical features")
+    with pytest.raises(ValueError, match=msg):
+        bin_mapper.transform(X, categorical_only=True)
+
+
+@pytest.mark.parametrize("n_bins", [7, 15, 256])
+def test_categorical_n_bins_greater_than_equal_cardinality(n_bins):
+    # test n_bins enough to hold all categories
+    X = np.array([[4] * 2 + [1] * 3 + [10] * 4 +
+                  [0] * 4 + [9] + [7] * 5], dtype=float).T
+
+    bin_mapper = _BinMapper(n_bins=n_bins, categorical=np.array([True])).fit(X)
+    assert bin_mapper.n_bins_non_missing_ == [6]
+    assert_allclose(bin_mapper.bin_categories_, [[0, 1, 4, 7, 9, 10]])
+
+    X_trans = bin_mapper.transform(
+        np.array([[10, 1, -1, 9, np.nan, 7, 4, 100, 0]], dtype=X_DTYPE).T)
+
+    # missing, negative, and unknown values are mapped to the missing bin
+    miss_bin = n_bins - 1
+    expected_trans = np.array([[5, 1, miss_bin, 4, miss_bin, 3, 2,
+                                miss_bin, 0]]).T
+    assert_array_equal(X_trans, expected_trans)
+
+
+@pytest.mark.parametrize(
+    "n_bins, expected_trans, bin_categories",
+    [
+        (3, [2, 2, 2, 2, 2, 1, 2, 2, 0], [0, 7]),
+        (4, [2, 3, 3, 3, 3, 1, 3, 3, 0], [0, 7, 10]),
+        (5, [3, 1, 4, 4, 4, 2, 4, 4, 0], [0, 1, 7, 10]),
+        (6, [4, 1, 5, 5, 5, 3, 2, 5, 0], [0, 1, 4, 7, 10]),
+    ])
+def test_categorical_n_bins_less_than_cardinality(
+        n_bins, expected_trans, bin_categories):
+    # test when there is not enough bins only the top n_bins - 1 distinct
+    # values (based on cardinality) will have its own bin
+    X = np.array([[4] * 2 + [1] * 3 + [10] * 4 +
+                  [0] * 4 + [9] + [7] * 5], dtype=X_DTYPE).T
+    X_test = np.array([[10, 1, -1, 9, np.nan, 7, 4, 100, 0]], dtype=X_DTYPE).T
+
+    bin_mapper = _BinMapper(n_bins=n_bins,
+                            categorical=np.array([True])).fit(X)
+    assert bin_mapper.n_bins_non_missing_ == [n_bins - 1]
+    assert_allclose(bin_mapper.bin_categories_, [bin_categories])
+    X_trans = bin_mapper.transform(X_test)
+
+    # missing, negative, unknown values are mapped to the missing bin
+    expected_trans = np.array([expected_trans]).T
+    assert_array_equal(X_trans, expected_trans)
+
+
+def test_categorical_n_bins_less_than_cardinality_ties():
```

Comment:
Looks like this is already implicitly covered in the test above, but can't hurt to check I guess